internal/node/generate_build_file.js

```diff
@@ -0,0 +1,166 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/**
+ * @fileoverview This script generates the BUILD.bazel file for NodeJS, npm & yarn.
+ */
+'use strict';
+
+const fs = require('fs');
+const path = require('path');
+
+const IS_WINDOWS = TEMPLATED_is_windows;
+const NODE_DIR = 'TEMPLATED_node_dir';
+const YARN_DIR = 'TEMPLATED_yarn_dir';
+const NODE_ACTUAL = 'TEMPLATED_node_actual';
+const NPM_ACTUAL = 'TEMPLATED_npm_actual';
+const YARN_ACTUAL = 'TEMPLATED_yarn_actual';
+
+if (require.main === module) {
+  main();
+}
+
+function mkdirp(dirname) {
+  if (!fs.existsSync(dirname)) {
+    mkdirp(path.dirname(dirname));
+    fs.mkdirSync(dirname);
+  }
+}
+
+function writeFileSync(filePath, contents) {
+  mkdirp(path.dirname(filePath));
+  fs.writeFileSync(filePath, contents);
+}
+
+/**
+ * Main entrypoint.
+ * Write BUILD file.
+ */
+function main() {
+  generateBuildFile()
+}
+
+module.exports = { main };
+
+function generateBuildFile() {
+  // *.pyc files are generated during node-gyp compilation and include
+  // absolute paths, making them non-hermetic.
+  // See https://github.com/bazelbuild/rules_nodejs/issues/347
+  const excludedFiles = ['.md', '.html', '.pyc'];
+  const nodejsSrcFiles = filterFilesForFilegroup(listFiles(NODE_DIR), [], excludedFiles)
+  const yarnSrcFiles = filterFilesForFilegroup(listFiles(YARN_DIR), [], excludedFiles)
+
+  const binaryExt = IS_WINDOWS ? '.cmd' : '';
+  const buildFile = `#Generated by node_repositories.bzl
+package(default_visibility = ["//visibility:public"])
+exports_files([
+  "run_npm.sh.template",
+  "bin/node_args.sh",
+  "${NODE_DIR}/bin/node",
+  "bin/node${binaryExt}",
+  ${!IS_WINDOWS ? '"bin/node.js",' : ''}
+  "bin/npm${binaryExt}",
+  "bin/npm_node_repositories${binaryExt}",
+  "bin/yarn${binaryExt}",
+  "bin/yarn_node_repositories${binaryExt}",
+  ])
+alias(name = "node", actual = "${NODE_ACTUAL}")
+alias(name = "npm", actual = "${NPM_ACTUAL}")
+alias(name = "yarn", actual = "${YARN_ACTUAL}")
+filegroup(
+  name = "node_runfiles",
+  srcs = [
+    ${!IS_WINDOWS ? '"bin/node.js",' : ''}
+    ${nodejsSrcFiles.map(f => `"${NODE_DIR}/${f}",`).join('\n    ')}
+    ${yarnSrcFiles.map(f => `"${YARN_DIR}/${f}",`).join('\n    ')}
+  ],
+)
+`
+  writeFileSync('BUILD.bazel', buildFile);
+}
+
+/**
+ * Returns an array of all the files under a directory as relative
+ * paths to the directory.
+ */
+function listFiles(rootDir, subDir = '') {
+  const dir = path.posix.join(rootDir, subDir);
+  return fs
+    .readdirSync(dir)
+    .reduce((files, file) => {
+      const fullPath = path.posix.join(dir, file);
+      const relPath = path.posix.join(subDir, file);
+      let stat;
+      try {
+        stat = fs.statSync(fullPath);
+      } catch (e) {
+        if (fs.lstatSync(fullPath).isSymbolicLink()) {
```

Comment:
I suspect this you won't run into symlinks here because you're just listing the contents of an extracted zip. This function could be simplified if you remove the check for broken symlinks and the check for BUILD files below.

Reply:
Done.